src/interface/swarm_pack.hpp

```diff
@@ -0,0 +1,207 @@
+
+  KOKKOS_INLINE_FUNCTION
+  SwarmPackIdx &operator=(std::size_t var_idx) {
+    vidx = var_idx;
+    offset = 0;
+    return *this;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  std::size_t VariableIdx() { return vidx; }
+  KOKKOS_INLINE_FUNCTION
+  int Offset() { return offset; }
+
+ private:
+  std::size_t vidx;
+  int offset;
+};
+
+// Operator overloads to make calls like `my_pack(b, my_pack_idx + 3, k, j, i)` work
+template <class T, REQUIRES(std::is_integral<T>::value)>
+KOKKOS_INLINE_FUNCTION SwarmPackIdx operator+(SwarmPackIdx idx, T offset) {
+  return SwarmPackIdx(idx.VariableIdx(), idx.Offset() + offset);
+}
+
+template <class T, REQUIRES(std::is_integral<T>::value)>
+KOKKOS_INLINE_FUNCTION SwarmPackIdx operator+(T offset, SwarmPackIdx idx) {
+  return idx + offset;
+}
+
+// Namespace in which to put variable name types that are used for indexing into
+// SwarmPack<[type list of variable name types]> on device
+namespace swarm_variable_names {
+// Struct that all variable_name types should inherit from
+template <bool REGEX, typename T = Real, int... NCOMP>
+struct base_t {
+  using data_type = T;
+
+  KOKKOS_INLINE_FUNCTION
+  base_t() : idx(0) {}
+
+  KOKKOS_INLINE_FUNCTION
+  explicit base_t(int idx1) : idx(idx1) {}
+
+  virtual ~base_t() = default;
+
+  // All of these are just static methods so that there is no
+  // extra storage in the struct
+  static std::string name() {
+    PARTHENON_FAIL("Need to implement your own name method.");
+    return "error";
+  }
+  KOKKOS_INLINE_FUNCTION
+  static bool regex() { return REGEX; }
+  KOKKOS_INLINE_FUNCTION
+  static int ndim() { return sizeof...(NCOMP); }
+  KOKKOS_INLINE_FUNCTION
+  static int size() { return multiply<NCOMP...>::value; }
+
+  const int idx;
+};
+
+} // namespace swarm_variable_names
+
+template <typename TYPE, class... Ts>
+class SwarmPack : public SwarmPackBase<TYPE> {
+ public:
+  using typename SwarmPackBase<TYPE>::pack_t;
+  using typename SwarmPackBase<TYPE>::desc_t;
+  using typename SwarmPackBase<TYPE>::bounds_t;
+  using typename SwarmPackBase<TYPE>::contexts_t;
+  using typename SwarmPackBase<TYPE>::max_active_indices_t;
+
+  using SwarmPackBase<TYPE>::pack_;
+  using SwarmPackBase<TYPE>::bounds_;
+  using SwarmPackBase<TYPE>::contexts_;
+  using SwarmPackBase<TYPE>::max_active_indices_;
+  using SwarmPackBase<TYPE>::nvar_;
+  using SwarmPackBase<TYPE>::nblocks_;
+
+  explicit SwarmPack(const SwarmPackBase<TYPE> &spb) : SwarmPackBase<TYPE>(spb) {
+    if constexpr (sizeof...(Ts) != 0) {
+      static_assert(std::is_same<TYPE, typename GetDataType<Ts...>::value>::value,
+                    "Type mismatch in SwarmPack! When passing type-based variables as "
+                    "template argument to SwarmPack, ensure that the first template "
+                    "parameter is a data type (e.g., Real or int) that matches the "
+                    "data type of subsequent variable types!");
+    }
```

Comment:
That should work since the right operand won't be evaluated when the left is `false`.

Reply:
hmm this didn't work... reverting for now... will look into more depth soon

Reply:
Is this assertion checked in a unit test? the argument of the `static_assert` is a bit opaque to me

Reply:
this assertion is not currently in a unit test (but I did test it downstream).  I am having a heck of a time getting unit tests to work with swarms because the unit tests are isolated (i.e., they don't follow the same initialization pathways a normal downstream code would see).